packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,172 @@
+          !ts.isConstructSignatureDeclaration(tsNode) &&
+          !ts.isMethodSignature(tsNode) &&
+          !ts.isIndexSignatureDeclaration(tsNode) &&
+          !ts.isFunctionTypeNode(tsNode) &&
+          !ts.isConstructorTypeNode(tsNode) &&
+          !ts.isJSDocFunctionType(tsNode) &&
+          !ts.isClassDeclaration(tsNode) &&
+          !ts.isClassExpression(tsNode)
+        ) {
+          declEndPos = tsNode.body?.getStart() ?? tsNode.end;
+        }
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const { uses } = usage.get(typeParameter.name)!;
+          let numExplicitUses = 0;
+          for (const use of uses) {
+            const pos = use.location.getStart();
+            if (pos > tsNode.getStart() && pos < declEndPos) {
+              numExplicitUses++;
+              if (numExplicitUses >= 2) {
+                break;
+              }
+            }
+          }
+          const inferredUses = inferredCounts?.get(typeParameter.name) ?? 0;
+          const numUses = numExplicitUses + inferredUses;
+          if (numUses === 1) {
+            context.report({
+              data: {
+                name: typeParameter.name.text,
+              },
+              node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+              messageId: 'sole',
+            });
+          }
+        }
+      },
+    };
+  },
+});
+
+// Fill in a counter of the number of times each type parameter appears in
+// the given type.
+function collectTypeParameterUsage(
+  checker: ts.TypeChecker,
+  rootType: ts.Type,
+  counts: Map<ts.Identifier, number>,
+): void {
+  const increment = (id: ts.Identifier): void => {
+    counts.set(id, 1 + (counts.get(id) ?? 0));
+  };
+
+  const process = (type: ts.Type): void => {
+    if (tsutils.isTypeParameter(type)) {
+      for (const decl of type.getSymbol()?.getDeclarations() ?? []) {
+        increment((decl as ts.TypeParameterDeclaration).name);
+        break;
+      }
+    }
+    // This should be "else if" but tsutils.isParameterType narrows type to never.
+    if (tsutils.isUnionOrIntersectionType(type)) {
+      type.types.forEach(process);
+    } else if (tsutils.isIndexedAccessType(type)) {
+      // This covers index access types like "T[K]"
+      process(type.objectType);
+      process(type.indexType);
+    } else if (tsutils.isTypeReference(type)) {
+      // This covers generic types like Map<K, V>.
+      for (const t of type.typeArguments ?? []) {
+        process(t);
+      }
+    } else if (tsutils.isObjectType(type)) {
+      // This covers inferred object types. This should be _after_
+      // isTypeReference to avoid descending into all the properties of a
+      // generic interface/class, e.g. Map<K, V>.
+      for (const sym of type.getProperties()) {
+        process(checker.getTypeOfSymbol(sym));
+      }
+    }
+    // TODO: mapped types, type predicate, conditional types, function types
```

Comment:
[Bug] One of the annoying bits of our general semver practices is that big changes to what a rule will or will not report on generally count as breaking changes. The rules are hand-wavy, but we've in the past found that newly reporting on large swathes of previously-allowed syntax can be received as a big breakage. We generally try to support them out-of-the-box.

Request: Can you include any test cases you're thinking of for them? Even if they're not passing right now, it'd be helpful to see.

Question: How terrible do you think it'd be to add support for these constructs?